Data/Text/ICU/Spoof.hsc

```diff
@@ -0,0 +1,273 @@
+  toBitMask RestrictionLevel = #const USPOOF_RESTRICTION_LEVEL
+  toBitMask Invisible = #const USPOOF_INVISIBLE
+  toBitMask CharLimit = #const USPOOF_CHAR_LIMIT
+  toBitMask MixedNumbers = #const USPOOF_MIXED_NUMBERS
+  toBitMask AllChecks = #const USPOOF_ALL_CHECKS
+  toBitMask AuxInfo = #const USPOOF_AUX_INFO
+
+type USpoofCheck = Int32
+
+data RestrictionLevel = ASCII
+                      | SingleScriptRestrictive
+                      | HighlyRestrictive
+                      | ModeratelyRestrictive
+                      | MinimallyRestrictive
+                      | Unrestrictive
+                      deriving (Bounded, Enum, Eq, Show)
+
+instance ToBitMask RestrictionLevel where
+  toBitMask ASCII = #const USPOOF_ASCII
+  toBitMask SingleScriptRestrictive = #const USPOOF_SINGLE_SCRIPT_RESTRICTIVE
+  toBitMask HighlyRestrictive = #const USPOOF_HIGHLY_RESTRICTIVE
+  toBitMask ModeratelyRestrictive = #const USPOOF_MODERATELY_RESTRICTIVE
+  toBitMask MinimallyRestrictive = #const USPOOF_MINIMALLY_RESTRICTIVE
+  toBitMask Unrestrictive = #const USPOOF_UNRESTRICTIVE
+
+type URestrictionLevel = Int32
+
+data SpoofCheckResult = CheckOK
+                      | CheckFailed [SpoofCheck]
+                      | CheckFailedWithRestrictionLevel { checks :: [SpoofCheck],
+                                                           level :: RestrictionLevel }
+                deriving (Eq, Show)
+
+data SkeletonTypeOverride = SkeletonSingleScript
+                          | SkeletonAnyCase
+                          deriving (Bounded, Enum, Eq, Show)
+
+instance ToBitMask SkeletonTypeOverride where
+  toBitMask SkeletonSingleScript = #const USPOOF_SINGLE_SCRIPT_CONFUSABLE
+  toBitMask SkeletonAnyCase = #const USPOOF_ANY_CASE
+
+type USkeletonTypeOverride = Int32
+
+makeSpoofCheckResult :: USpoofCheck -> SpoofCheckResult
+makeSpoofCheckResult c =
+  case spoofChecks of
+    [] -> CheckOK
+    _ ->
+      case restrictionLevel of
+        Nothing -> CheckFailed spoofChecks
+        Just l -> CheckFailedWithRestrictionLevel spoofChecks l
+  where spoofChecks = fromBitMask $ fromIntegral $ c .&. #const USPOOF_ALL_CHECKS
+        restrictionValue = c .&. #const USPOOF_RESTRICTION_LEVEL_MASK
+        restrictionLevel = highestValueInBitMask $ fromIntegral $ restrictionValue
+
+-- | Open a spoof checker for checking Unicode strings for lookalike security issues.
+open :: IO MSpoof
+open = wrap =<< handleError uspoof_open
+
+-- | Open a spoof checker previously serialized to bytes using 'serialize'.
+-- The ForeignPtr in the ByteString will remain referenced for the
+-- lifetime of the returned object. The memory it points to must not
+-- be modified until the returned object is closed.
+openFromSerialized :: ByteString -> IO MSpoof
+openFromSerialized b =
+  case toForeignPtr b of
+    (ptr, off, len) -> withForeignPtr ptr $ \p ->
+      wrapWithSerialized ptr =<< handleError (uspoof_openFromSerialized (p `plusPtr` off) (fromIntegral len) nullPtr)
```

Comment:
This API looks wrong to me. The `MSpoof` type should retain a reference to the `ForeignPtr` inside the `ByteString` as long as the `MSpoof` is open so that there is no chance of a runtime crash due to the `ForeignPtr` being garbage collected while the `MSpoof` is still alive. Putting that onus on the caller is not the right call.